middleware/checkAuth.js

```diff
@@ -0,0 +1,85 @@
+const jwt = require('jsonwebtoken');
+const ERROR = require('../constants').ERROR;
+const User = require('../models/User');
+
+const TOKEN_EXPIRED_ERROR = 'TokenExpiredError';
+
+module.exports = async (req, res, next) => {
+  if (req.method === 'OPTIONS') {
+    return next();
+  }
+
+  let user;
+  let accessToken;
+  let decodedAccessToken;
+  let refreshToken;
+  let decodedRefreshToken;
+
+  try {
+    accessToken = req.headers.authorization.split(' ')[1];
+
+    decodedAccessToken = jwt.verify(accessToken, process.env.JWT_SECRET_KEY);
+
+    const { id, nickname, imageUrl } = decodedAccessToken;
+
+    user = await User.findById(id);
+
+    refreshToken = user.refreshToken;
+
+    try {
+      decodedRefreshToken = jwt.verify(user.refreshToken, process.env.JWT_SECRET_KEY);
+
+    } catch (error) {
+      if (error.name = TOKEN_EXPIRED_ERROR) {
+        const newRefreshToken = jwt.sign({
+          id,
+          nickname,
+          imageUrl,
+        }, process.env.JWT_ACCESS_TOKEN_EXPIRES_IN, {
+          expiresIn: Number(process.env.JWT_REFRESH_TOKEN_EXPIRES_IN)
+        });
+
+        user.refreshToken = newRefreshToken;
+        await user.save();
+        req.user = { id, nickname, imageUrl };
+        next();
+      }
+    }
+
+    req.user = { id, nickname, imageUrl };
+    next();
+  } catch (error) {
+
+    if (error.name = TOKEN_EXPIRED_ERROR) {
+      decoded = jwt.verify(accessToken, process.env.JWT_SECRET_KEY, { ignoreExpiration: true });
+
+      const { id, nickname, imageUrl } = decodedAccessToken;
+
+      try {
+        decodedRefreshToken = jwt.verify(user.refreshToken, process.env.JWT_SECRET_KEY);
+
+      } catch (error) {
+        const err = new Error(ERROR.AUTHENTICATION_FAILED);
+
+        err.status = 401;
+        return next(err);
```

Comment:
에러 메세지를 그대로 노출 시키지 않기 위해 새로운 에러 객체를 만들어주신걸까요? 저희 앱의 에러 핸들러 설정상 NODE_ENV 환경변수가 'development'일때만 에러 메세지를 클라이언트에 보내주므로, 캐치한 에러를 next로 그대로 넘겨줘도 될 것 같은데, 어떻게 생각하시는지요?
```suggestion
      } catch (error) {
        error.status = 401;
        return next(error);
```

참고) `req.app.get('env')`에 대해
https://stackoverflow.com/questions/34227216/process-env-vs-app-getenv-on-getting-the-express-js-environment

Reply:
제가 이 부분에서 고려했던 부분은 직접적으로 사용자가 사용 할때도 해당 메시지 부분을 볼 수 있다는 걸 고려한 측면에서 커스텀 해서 넘겨 주었던 부분인데 제가 현재 저희 앱에 백앤드 환경 설정 부분을 간과했나 보네요.. 이 부분은 혜민님의 피드백에 따라 status 코드만 찍어주고 넘겨주는 방식으로 진행하겠습니다.